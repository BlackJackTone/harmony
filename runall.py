import subprocess
import time
import json
import os.path

tests = [
    { "args": "code/triangle.hny", "issue": "No issues", "nstates": 13 },
    { "args": "code/prog1.hny", "issue": "No issues", "nstates": 2 },
    { "args": "code/prog2.hny", "issue": "Safety violation", "nstates": 11 },
    { "args": "code/Up.hny", "issue": "Safety violation", "nstates": 44 },
    { "args": "code/Upf.hny", "issue": "Finally predicate violation", "nstates": 13 },
    { "args": "code/UpEnter.hny", "issue": "Non-terminating state", "nstates": 59 },
    { "args": "code/csbarebones.hny", "issue": "Safety violation", "nstates": 5 },
    { "args": "code/cs.hny", "issue": "Safety violation", "nstates": 11 },
    { "args": "code/naiveLock.hny", "issue": "Safety violation", "nstates": 34 },
    { "args": "code/naiveFlags.hny", "issue": "Non-terminating state", "nstates": 45 },
    { "args": "code/naiveTurn.hny", "issue": "Non-terminating state", "nstates": 28 },
    { "args": "code/Peterson.hny", "issue": "No issues", "nstates": 104 },
    { "args": "code/PetersonBroken.hny", "issue": "Safety violation", "nstates": 126 },
    { "args": "code/PetersonInductive.hny", "issue": "No issues", "nstates": 104 },
    { "args": "code/csonebit.hny", "issue": "Active busy waiting", "nstates": 73 },
    { "args": "code/PetersonMethod.hny", "issue": "No issues", "nstates": 104 },
    { "args": "code/hanoi.hny", "issue": "Safety violation", "nstates": 28 },
    { "args": "code/clock.hny", "issue": "Safety violation", "nstates": 5462 },
    { "args": "code/cssynch.hny", "issue": "No issues", "nstates": 35 },
    { "args": "-msynch=synchS code/cssynch.hny", "issue": "No issues", "nstates": 58 },
    { "args": "-msynch=ticket code/cssynch.hny", "issue": "No issues", "nstates": 458 },
    { "args": "code/UpLock.hny", "issue": "No issues", "nstates": 48 },
    { "args": "-msynch=synchS code/UpLock.hny", "issue": "No issues", "nstates": 58 },
    { "args": "code/spinlock.hny", "issue": "No issues", "nstates": 473 },
    { "args": "code/xy.hny", "issue": "Safety violation", "nstates": 19 },
    { "args": "code/atm.hny", "issue": "Invariant violation", "nstates": 1832 },
    { "args": "code/queuedemo.hny", "issue": "No issues", "nstates": 80 },
    { "args": "code/setobjtest.hny", "issue": "No issues", "nstates": 217 },
    { "args": "-msetobj=linkedlist code/setobjtest.hny", "issue": "No issues", "nstates": 9495 },
    { "args": "code/qtestseq.hny", "issue": "No issues", "nstates": 600 },
    { "args": "-mqueueconc=queueMS code/qtestseq.hny", "issue": "No issues", "nstates": 702 },
    { "args": "code/qtest1.hny", "issue": "No issues", "nstates": 202 },
    { "args": "code/qtest2.hny", "issue": "No issues", "nstates": 55 },
    { "args": "code/qtest3.hny", "issue": "No issues", "nstates": 3225 },
    { "args": "code/qtest4.hny", "issue": "No issues", "nstates": 10 },
    { "args": "code/qtestconc.hny", "issue": "No issues", "nstates": 1879206 },
    { "args": "-msynch=synchS code/qtestconc.hny", "issue": "No issues", "nstates": 3430951 },
    { "args": "-o queue4.hfa code/qtestpar.hny", "issue": "No issues", "nstates": 3385 },
    { "args": "-B queue4.hfa -m queue=queueconc code/qtestpar.hny", "issue": "No issues", "nstates": 66530 },
    { "args": "-o queue4.hfa code/qtestpar.hny", "issue": "No issues", "nstates": 3385 },
    { "args": "-B queue4.hfa -m queue=queueMS code/qtestpar.hny", "issue": "No issues", "nstates": 99816 },
    { "args": "-mqueue=queuebroken code/qtestpar.hny", "issue": "Safety violation", "nstates": 5 },
    { "args": "code/RWtest.hny", "issue": "No issues", "nstates": 135 },
    { "args": "-mRW=RWsbs code/RWtest.hny", "issue": "No issues", "nstates": 1757 },
    { "args": "-mRW=RWcv code/RWtest.hny", "issue": "No issues", "nstates": 1691 },
    { "args": "-mRW=RWcv -msynch=synchS code/RWtest.hny", "issue": "No issues", "nstates": 4301 },
    { "args": "-mRW=RWfair code/RWtest.hny", "issue": "No issues", "nstates": 1248 },
    { "args": "-mRW=RWfair -msynch=synchS code/RWtest.hny", "issue": "No issues", "nstates": 3137 },
    { "args": "-o rw.hfa -cNOPS=2 code/RWbtest.hny", "issue": "No issues", "nstates": 273 },
    { "args": "-B rw.hfa -cNOPS=2 -m RW=RWsbs code/RWbtest.hny", "issue": "No issues", "nstates": 1393 },
    { "args": "-o rw.hfa -cNOPS=2 code/RWbtest.hny", "issue": "No issues", "nstates": 273 },
    { "args": "-B rw.hfa -cNOPS=2 -m RW=RWcv code/RWbtest.hny", "issue": "No issues", "nstates": 923 },
    { "args": "-B rw.hfa -cNOPS=2 -m RW=RWfair code/RWbtest.hny", "issue": "No issues", "nstates": 1019 },
    { "args": "-B rw.hfa -cNOPS=2 -m RW=RWcheat code/RWbtest.hny", "issue": "No issues", "nstates": 257 },
    { "args": "-mboundedbuffer=BBhoare code/BBtest.hny", "issue": "No issues", "nstates": 1017 },
    { "args": "-mboundedbuffer=BBhoare -msynch=synchS code/BBtest.hny", "issue": "No issues", "nstates": 4019 },
    { "args": "code/qsorttest.hny", "issue": "No issues", "nstates": 1190 },
    { "args": "code/Diners.hny", "issue": "Non-terminating state", "nstates": 9095 },
#    { "args": "-msynch=synchS code/Diners.hny", "issue": "Non-terminating state", "nstates": 32476 },
    { "args": "code/DinersCV.hny", "issue": "No issues", "nstates": 101549 },
    { "args": "-msynch=synchS code/DinersCV.hny", "issue": "No issues", "nstates": 2293519 },
    { "args": "code/DinersAvoid.hny", "issue": "No issues", "nstates": 39884 },
    { "args": "-msynch=synchS code/DinersAvoid.hny", "issue": "No issues", "nstates": 152874 },
    { "args": "code/bank.hny", "issue": "Non-terminating state", "nstates": 2827 },
    { "args": "code/counter.hny", "issue": "No issues", "nstates": 508 },
    { "args": "code/qbarrier.hny", "issue": "No issues", "nstates": 1664 },
    { "args": "-msynch=synchS code/qbarrier.hny", "issue": "No issues", "nstates": 4927 },
    { "args": "code/barriertest.hny", "issue": "No issues", "nstates": 2529 },
    { "args": "-msynch=synchS code/barriertest.hny", "issue": "No issues", "nstates": 7425 },
    { "args": "code/barriertest2.hny", "issue": "No issues", "nstates": 5711 },
    { "args": "-msynch=synchS code/barriertest2.hny", "issue": "No issues", "nstates": 16511 },
    { "args": "code/trap.hny", "issue": "No issues", "nstates": 8 },
    { "args": "code/trap2.hny", "issue": "Finally predicate violation", "nstates": 21 },
    { "args": "code/trap3.hny", "issue": "Non-terminating state", "nstates": 26 },
    { "args": "code/trap4.hny", "issue": "No issues", "nstates": 16 },
    { "args": "code/trap5.hny", "issue": "No issues", "nstates": 16 },
    { "args": "code/trap6.hny", "issue": "No issues", "nstates": 386 },
    { "args": "-msynch=synchS code/trap6.hny", "issue": "No issues", "nstates": 554 },
    { "args": "code/hw.hny", "issue": "No issues", "nstates": 23864 },
    { "args": "code/abptest.hny", "issue": "No issues", "nstates": 2778 },
    { "args": "code/leader.hny", "issue": "No issues", "nstates": 33005 },
    { "args": "-mstack=stack1 code/stacktest.hny", "issue": "No issues", "nstates": 2 },
    { "args": "-mstack=stack2 code/stacktest.hny", "issue": "No issues", "nstates": 2 },
    { "args": "-mstack=stack3 code/stacktest.hny", "issue": "No issues", "nstates": 2 },
    { "args": "-mstack=stack4 code/stacktest.hny", "issue": "No issues", "nstates": 2 }
]

for t in tests:
    print()
    print("Running harmony %s"%t["args"], flush=True)
    start = time.time()
    args = t["args"].split(" ")
    cp = subprocess.run(["./harmony", "--noweb"] + args, capture_output=True)
    end = time.time()
    if cp.returncode == 0:
        (base, ext) = os.path.splitext(args[-1])
        with open(base + ".hco") as f:
            hco = json.load(f)
            print("time=%.2f, #states=%d, issue=%s" % (end - start, hco["nstates"], hco["issue"]))
            if t["issue"] != hco["issue"]:
                print("Different issue (was %s)???  Aborting further tests" % t["issue"])
                break
            if t["issue"] != "Safety violation" and min(t["nstates"], hco["nstates"]) / max(t["nstates"], hco["nstates"]) < .9:
                print("#states very different (was %d)???  Aborting further tests" % t["nstates"])
                break
    else:
        print("Error code %d, aborting further tests" % cp.returncode)
        print("Output: ", cp.stdout)
        print("Error: ", cp.stderr)
        break
