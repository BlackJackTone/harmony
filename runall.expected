==============
prog1
==============
#states 2
2 components, 0 bad states
No issues
==============
prog2
==============
#states 10
Safety Violation
T0: __init__() [0-3,17-25] { shared: True }
T1: g() [13-16] { shared: False }
T2: f() [4-8] { shared: False }
Harmony assertion failed
==============
Up
==============
#states 44
Safety Violation
T0: __init__() [0-5,33-41] { count: 0, done: [ False, False ] }
T1: incrementer(0) [6-9] { count: 0, done: [ False, False ] }
T2: incrementer(1) [6-20] { count: 1, done: [ False, True ] }
T1: incrementer(0) [10-29] { count: 1, done: [ True, True ] }
Harmony assertion failed
==============
UpEnter
==============
#states 60
60 components, 1 bad states
Non-terminating state
T0: __init__() [0-7,55-63] { count: 0, done: [ False, False ], entered: [ False, False ] }
T1: incrementer(0) [8-18] { count: 0, done: [ False, False ], entered: [ True, False ] }
T2: incrementer(1) [8-25] { count: 0, done: [ False, False ], entered: [ True, True ] }
T1: incrementer(0) [19-25] { count: 0, done: [ False, False ], entered: [ True, True ] }

==============
csbarebones
==============
#states 7
Safety Violation
T0: __init__() [0,1,28-36] { }
T1: thread(0) [2-4] { }
T2: thread(1) [2-21] { }
Harmony assertion failed
==============
cs
==============
#states 17
Safety Violation
T0: __init__() [0,1,29-37] { }
T1: thread(0) [2-4(choose True),5] { }
T2: thread(1) [2-4(choose True),5-22] { }
Harmony assertion failed
==============
naiveLock
==============
#states 49
Safety Violation
T0: __init__() [0-3,37-45] { lockTaken: False }
T1: thread(0) [4-6(choose True),7-10] { lockTaken: False }
T2: thread(1) [4-6(choose True),7-11] { lockTaken: True }
T1: thread(0) [11-28] { lockTaken: True }
Harmony assertion failed
==============
naiveFlags
==============
#states 45
15 components, 1 bad states
Non-terminating state
T0: __init__() [0-3,48-56] { flags: [ False, False ] }
T1: thread(0) [4-6(choose True),7-17] { flags: [ True, False ] }
T2: thread(1) [4-6(choose True),7-17] { flags: [ True, True ] }

==============
naiveTurn
==============
#states 28
21 components, 2 bad states
Non-terminating state
T0: __init__() [0-3,39-47] { turn: 0 }
T1: thread(0) [4-6(choose True),7-11] { turn: 1 }
T2: thread(1) [4-6(choose False),7,37,38] { turn: 1 }

==============
Peterson
==============
#states 104
37 components, 0 bad states
No issues
==============
PetersonInductive
==============
#states 104
37 components, 0 bad states
No issues
==============
csonebit
==============
#states 73
18 components, 0 bad states
Active busy waiting
T0: __init__() [0-5,61-69] { flags: [ False, False ] }
T1: thread(0) [6-8(choose True),9-19] { flags: [ True, False ] }
T2: thread(1) [6-8(choose True),9-19] { flags: [ True, True ] }

==============
PetersonMethod
==============
#states 104
37 components, 0 bad states
No issues
==============
clock
==============
#states 36894
Safety Violation
T0: __init__() [0,1,134-141,2-21,142-146,2-21,147-174,181-197,23-45,81-133,198-207,23-45,81-133,208-228,163-174,181-197,23-45,81-133,198-207,23-45,81-133,208-228,163-174,181-197,23-45,81-133,198-207,23-45,81-133,208-228,163-174,181-197,23-80,31-80,31-80,31-45,81-133,198-207,23-45,81-133,208-228,163-170,175-180(choose 2),181-197,23-30,116-133,198-207,23-30,116-133,208-228,163-170,175-180(choose 1),181-197,23-80,31-45,81-133,198-207,23-30,116-133,208-228,163-170,175-180(choose 2),181-197,23-30,116-133,198-207,23-30,116-133,208-228,163-170,175-180(choose 5),181-197,23-80,31-80,31-80,31-45,81-133,198-207,23-80,31-80,31-80,31-80,31-45,81-133,208-228,163-170,175-180(choose 1),181-197,23-30,116-133,198-207,23-45,81-133,208-228,163-170,175-180(choose 2),181-197,23-30,116-133,198-207,23-45,81-133,208-220] { clock3: { .entries: [ 5, 2, 1 ], .hand: 1, .misses: 6, .recent: { 1, 2, 5 } }, clock4: { .entries: [ 5, 1, 2, 4 ], .hand: 3, .misses: 7, .recent: { 1, 2, 5 } }, refs: [ 1, 2, 3, 4, 2, 1, 2, 5, 1, 2 ] }
Harmony assertion failed
==============
spinlock
==============
#states 401
124 components, 0 bad states
No issues
==============
csTAS
==============
#states 322
157 components, 0 bad states
No issues
==============
UpLock
==============
#states 49
49 components, 0 bad states
No issues
==============
UpLock synchS
==============
#states 59
59 components, 0 bad states
No issues
==============
xy
==============
#states 19
Safety Violation
T0: __init__() [0-9,54-62] { x: 0, y: 100 }
T1: setX(50) [10-16] { x: 50, y: 100 }
T2: checker() [32-35,20-30,36-50] { x: 50, y: 100 }
Harmony assertion failed: (50, 100)
==============
atm
==============
#states 2885
Invariant Violation
T0: __init__() [0-5,369-371,1047-1062,769-773,762-767,774,775,1063-1066(choose 0),1067-1070,1052-1062,769-773,762-767,774,775,1063-1066(choose 1),1067-1070,1052-1054,1071-1074,1101,1191-1205(choose 1),1206-1209(choose 1),1210-1213,1193-1205(choose 1),1206-1209(choose 1),1210-1213,1193-1195,1214,1215] { accounts: [ { .balance: 0, .lock: False }, { .balance: 1, .lock: False } ], bag: (), list: (), synch: () }
T2: customer(2, 1, 1) [1164-1168,1102-1110,785-788,727-738,789-791,1111-1126,793-807,1127,1128,1169-1185,1130-1138,785-788,727-730] { accounts: [ { .balance: 0, .lock: False }, { .balance: 1, .lock: False } ], bag: (), list: (), synch: () }
T1: customer(1, 1, 1) [1164-1168,1102-1110,785-788,727-738,789-791,1111-1126,793-807,1127,1128,1169-1185,1130-1138,785-788,727-738,789-791,1139-1160,793-807,1161,1162,1186-1190] { accounts: [ { .balance: 0, .lock: False }, { .balance: 0, .lock: False } ], bag: (), list: (), synch: () }
T2: customer(2, 1, 1) [731-738,789-791,1139-1160,793-797] { accounts: [ { .balance: 0, .lock: False }, { .balance: -1, .lock: True } ], bag: (), list: (), synch: () }

==============
queuedemo
==============
#states 912
912 components, 0 bad states
No issues
==============
queuedemo synchS
==============
#states 1998
1998 components, 0 bad states
No issues
==============
qtestseq
==============
#states 600
600 components, 0 bad states
No issues
==============
qtest1
==============
#states 5058
5058 components, 0 bad states
No issues
==============
qtest1 synchS
==============
#states 7218
7218 components, 0 bad states
No issues
==============
qtest2
==============
#states 219
219 components, 0 bad states
No issues
==============
qtest2 synchS
==============
#states 309
309 components, 0 bad states
No issues
==============
qtest3
==============
#states 336133
331885 components, 0 bad states
No issues
==============
qtest3 synchS
==============
#states 875939
869801 components, 0 bad states
No issues
==============
qtest4
==============
#states 122
122 components, 0 bad states
No issues
==============
qtest4 synchS
==============
#states 122
122 components, 0 bad states
No issues
==============
qtestconc.hny
==============
#states 1879206
1879206 components, 0 bad states
No issues
==============
qtestconc.hny synchS
==============
#states 3430951
3430951 components, 0 bad states
No issues
==============
queueMS.hny
==============
#states 2095
2095 components, 0 bad states
Data race (?alloc.pool[0].next)
T0: __init__() [0-7,371-373,1049-1055,1254-1256,1096-1100,1056-1074,1101-1113,771-775,764-769,776,777,1114-1118,771-775,764-769,776,777,1119-1121,1257-1286] { alloc: { .next: 1, .pool: [ { .next: None, .value: () } ] }, bag: (), demoq: { .hdlock: False, .head: ?alloc.pool[0], .tail: ?alloc.pool[0], .tllock: False }, list: (), queue: (), synch: () }
T3: sender(?demoq, 1) [1223-1234,1123-1133,1056-1074,1134-1139,787-790,729-740,791-793,1140-1147] { alloc: { .next: 2, .pool: [ { .next: None, .value: () }, { .next: None, .value: 1 } ] }, bag: (), demoq: { .hdlock: False, .head: ?alloc.pool[0], .tail: ?alloc.pool[0], .tllock: True }, list: (), queue: (), synch: () }
T1: receiver(?demoq) [1238-1242,1164-1170,787-790,729-740,791-793,1171-1179] { alloc: { .next: 2, .pool: [ { .next: None, .value: () }, { .next: None, .value: 1 } ] }, bag: (), demoq: { .hdlock: True, .head: ?alloc.pool[0], .tail: ?alloc.pool[0], .tllock: True }, list: (), queue: (), synch: () }

==============
queueMS.hny synchS
==============
#states 3287
3287 components, 0 bad states
Data race (?alloc.pool[0].next)
T0: __init__() [0-7,805-811,1010-1012,852-856,812-830,857-869,420-424,407-418,425,426,870-874,420-424,407-418,425,426,875-877,1013-1042] { alloc: { .next: 1, .pool: [ { .next: None, .value: () } ] }, demoq: { .hdlock: { .acquired: False, .suspended: () }, .head: ?alloc.pool[0], .tail: ?alloc.pool[0], .tllock: { .acquired: False, .suspended: () } }, list: (), queue: (), synch: () }
T1: sender(?demoq, 1) [979-990,879-889,812-830,890-895,428-434,458-465,896-903] { alloc: { .next: 2, .pool: [ { .next: None, .value: () }, { .next: None, .value: 1 } ] }, demoq: { .hdlock: { .acquired: False, .suspended: () }, .head: ?alloc.pool[0], .tail: ?alloc.pool[0], .tllock: { .acquired: True, .suspended: () } }, list: (), queue: (), synch: () }
T3: receiver(?demoq) [994-998,920-926,428-434,458-465,927-935] { alloc: { .next: 2, .pool: [ { .next: None, .value: () }, { .next: None, .value: 1 } ] }, demoq: { .hdlock: { .acquired: True, .suspended: () }, .head: ?alloc.pool[0], .tail: ?alloc.pool[0], .tllock: { .acquired: True, .suspended: () } }, list: (), queue: (), synch: () }

==============
lltest
==============
#states 9498
9498 components, 0 bad states
No issues
==============
lltest synchS
==============
#states 18377
18377 components, 0 bad states
No issues
==============
RWtest
==============
#states 949
258 components, 0 bad states
Active busy waiting
T0: __init__() [0-7,371-373,1201-1203,1050-1056,771-775,764-769,776,777,1057-1065,1204,1205,1263-1274,1265-1274,1265-1274,1265-1267,1275,1276] { RW: (), bag: (), list: (), rw: { .lock: False, .nreaders: 0, .nwriters: 0 }, synch: () }
T1: thread() [1206-1208(choose True),1209-1211(choose .read),1212-1217,1067-1072,787-790,729-740,791-793,1073-1080,1094-1107,795-809,1108,1109,1218] { RW: (), bag: (), list: (), rw: { .lock: False, .nreaders: 1, .nwriters: 0 }, synch: () }
T2: thread() [1206-1208(choose True),1209-1211(choose .write),1212-1214,1235-1237,1135-1140,787-790,729-740,791-793,1141-1144] { RW: (), bag: (), list: (), rw: { .lock: True, .nreaders: 1, .nwriters: 0 }, synch: () }

==============
RWtest synchS
==============
#states 2735
452 components, 0 bad states
Active busy waiting
T0: __init__() [0-7,957-959,806-812,420-424,407-418,425,426,813-821,960,961,1019-1030,1021-1030,1021-1030,1021-1023,1031,1032] { RW: (), list: (), rw: { .lock: { .acquired: False, .suspended: () }, .nreaders: 0, .nwriters: 0 }, synch: () }
T1: thread() [962-964(choose True),965-967(choose .read),968-973,823-828,428-434,458-465,829-836,850-863,467-491,511,512,864,865,974] { RW: (), list: (), rw: { .lock: { .acquired: False, .suspended: () }, .nreaders: 1, .nwriters: 0 }, synch: () }
T2: thread() [962-964(choose True),965-967(choose .write),968-970,991-993,891-896,428-434,458-465,897-900] { RW: (), list: (), rw: { .lock: { .acquired: True, .suspended: () }, .nreaders: 1, .nwriters: 0 }, synch: () }

==============
RWsbs
==============
#states 1757
558 components, 0 bad states
No issues
==============
RWsbs synchS
==============
#states 4212
897 components, 0 bad states
No issues
==============
RWfair
==============
#states 1248
387 components, 0 bad states
No issues
==============
RWfair synchS
==============
#states 3137
635 components, 0 bad states
No issues
==============
BBhoaretest
==============
#states 1045
1045 components, 0 bad states
No issues
==============
BBhoaretest synchS
==============
#states 4127
4127 components, 0 bad states
No issues
==============
RWcv
==============
#states 1673
349 components, 0 bad states
No issues
==============
RWcv synchS
==============
#states 4301
666 components, 0 bad states
No issues
==============
qsorttest
==============
#states 454
454 components, 0 bad states
No issues
==============
Diners
==============
#states 9095
1700 components, 1 bad states
Non-terminating state
T0: __init__() [0-5,369-371,1047-1051,769-773,762-767,774,775,1052-1056,1102-1113,1104-1113,1104-1113,1104-1113,1104-1113,1104-1106,1114,1115] { bag: (), forks: [ False, False, False, False, False ], list: (), synch: () }
T1: diner(0) [1057-1072(choose True),1073-1078,785-788,727-738,789-791,1079-1084,785-788,727-730] { bag: (), forks: [ True, False, False, False, False ], list: (), synch: () }
T2: diner(1) [1057-1072(choose True),1073-1078,785-788,727-738,789-791,1079-1084,785-788,727-730] { bag: (), forks: [ True, True, False, False, False ], list: (), synch: () }
T3: diner(2) [1057-1072(choose True),1073-1078,785-788,727-738,789-791,1079-1084,785-788,727-730] { bag: (), forks: [ True, True, True, False, False ], list: (), synch: () }
T4: diner(3) [1057-1072(choose True),1073-1078,785-788,727-738,789-791,1079-1084,785-788,727-730] { bag: (), forks: [ True, True, True, True, False ], list: (), synch: () }
T5: diner(4) [1057-1072(choose True),1073-1078,785-788,727-738,789-791,1079-1084,785-788,727-730] { bag: (), forks: [ True, True, True, True, True ], list: (), synch: () }

==============
Diners synchS
==============
#states 32476
6416 components, 1 bad states
Non-terminating state
T0: __init__() [0-5,803-807,418-422,405-416,423,424,808-812,858-869,860-869,860-869,860-869,860-869,860-862,870,871] { forks: [ { .acquired: False, .suspended: () }, { .acquired: False, .suspended: () }, { .acquired: False, .suspended: () }, { .acquired: False, .suspended: () }, { .acquired: False, .suspended: () } ], list: (), synch: () }
T1: diner(0) [813-828(choose True),829-834,426-432,456-463,835-840,426-430] { forks: [ { .acquired: True, .suspended: () }, { .acquired: False, .suspended: () }, { .acquired: False, .suspended: () }, { .acquired: False, .suspended: () }, { .acquired: False, .suspended: () } ], list: (), synch: () }
T5: diner(4) [813-828(choose True),829-834,426-432,456-463,835-840,426-442] { forks: [ { .acquired: True, .suspended: [ CONTEXT(.diner) ] }, { .acquired: False, .suspended: () }, { .acquired: False, .suspended: () }, { .acquired: False, .suspended: () }, { .acquired: True, .suspended: () } ], list: (), synch: () }
T4: diner(3) [813-828(choose True),829-834,426-432,456-463,835-840,426-442] { forks: [ { .acquired: True, .suspended: [ CONTEXT(.diner) ] }, { .acquired: False, .suspended: () }, { .acquired: False, .suspended: () }, { .acquired: True, .suspended: () }, { .acquired: True, .suspended: [ CONTEXT(.diner) ] } ], list: (), synch: () }
T3: diner(2) [813-828(choose True),829-834,426-432,456-463,835-840,426-442] { forks: [ { .acquired: True, .suspended: [ CONTEXT(.diner) ] }, { .acquired: False, .suspended: () }, { .acquired: True, .suspended: () }, { .acquired: True, .suspended: [ CONTEXT(.diner) ] }, { .acquired: True, .suspended: [ CONTEXT(.diner) ] } ], list: (), synch: () }
T2: diner(1) [813-828(choose True),829-834,426-432,456-463,835-840,426-442] { forks: [ { .acquired: True, .suspended: [ CONTEXT(.diner) ] }, { .acquired: True, .suspended: () }, { .acquired: True, .suspended: [ CONTEXT(.diner) ] }, { .acquired: True, .suspended: [ CONTEXT(.diner) ] }, { .acquired: True, .suspended: [ CONTEXT(.diner) ] } ], list: (), synch: () }
T1: diner(0) [431-442] { forks: [ { .acquired: True, .suspended: [ CONTEXT(.diner) ] }, { .acquired: True, .suspended: [ CONTEXT(.diner) ] }, { .acquired: True, .suspended: [ CONTEXT(.diner) ] }, { .acquired: True, .suspended: [ CONTEXT(.diner) ] }, { .acquired: True, .suspended: [ CONTEXT(.diner) ] } ], list: (), synch: () }

==============
DinersCV
==============
#states 99789
34734 components, 0 bad states
No issues
==============
DinersCV synchS
==============
#states 2293519
441764 components, 0 bad states
No issues
==============
DinersAvoid
==============
#states 39884
2129 components, 0 bad states
No issues
==============
DinersAvoid synchS
==============
#states 152874
10504 components, 0 bad states
No issues
==============
bank
==============
#states 2582
2582 components, 9 bad states
Non-terminating state
T0: __init__() [0-5,369-371,1047-1062,769-773,762-767,774,775,1063-1066(choose 1),1067-1070,1052-1062,769-773,762-767,774,775,1063-1066(choose 1),1067-1070,1052-1054,1071-1074,1178-1189,1180-1189,1180-1182,1190,1191] { accounts: [ { .balance: 1, .lock: False }, { .balance: 1, .lock: False } ], bag: (), list: (), synch: () }
T1: thread() [1150-1152(choose 0),1153-1159(choose 1),1160-1173(choose 1),1174,1175,1075-1083,785-788,727-738,789-791,1084-1110,785-788,727-730] { accounts: [ { .balance: 0, .lock: True }, { .balance: 1, .lock: False } ], bag: (), list: (), synch: () }
T2: thread() [1150-1152(choose 1),1153-1159(choose 0),1160-1173(choose 1),1174,1175,1075-1083,785-788,727-738,789-791,1084-1110,785-788,727-730] { accounts: [ { .balance: 0, .lock: True }, { .balance: 0, .lock: True } ], bag: (), list: (), synch: () }

==============
counter
==============
#states 620
620 components, 0 bad states
No issues
==============
qbarrier
==============
#states 2338
2338 components, 0 bad states
No issues
==============
qbarrier synchS
==============
#states 6285
6285 components, 0 bad states
No issues
==============
barriertest
==============
#states 6135
6135 components, 0 bad states
No issues
==============
barriertest synchS
==============
#states 18741
18741 components, 0 bad states
No issues
==============
trap
==============
#states 9
9 components, 0 bad states
No issues
==============
trap2
==============
#states 23
Safety Violation
T0: __init__() [0-7,36-40] { count: 0, done: False }
T1: main() [17-24,interrupt,8-15,24-32] { count: 1, done: True }
Harmony assertion failed
==============
trap3
==============
#states 11
Safety Violation
T0: __init__() [0-5,369-371,1047-1051,769-773,762-767,774,775,1052-1057,1102-1106] { bag: (), count: 0, countlock: False, done: False, list: (), synch: () }
T1: main() [1075-1081,785-788,727-738,789-791,1082-1089,793-807,1090] { bag: (), count: 1, countlock: False, done: False, list: (), synch: () }

==============
trap4
==============
#states 11
11 components, 0 bad states
No issues
==============
trap5
==============
#states 11
11 components, 0 bad states
No issues
==============
trap6
==============
#states 117
117 components, 0 bad states
No issues
==============
trap6 synchS
==============
#states 161
161 components, 0 bad states
No issues
==============
hw
==============
#states 15737
13748 components, 0 bad states
No issues
==============
abptest
==============
#states 1536
608 components, 0 bad states
No issues
